### PR TITLE
Server should send 2 goaway messages to gracefully shutdown the connection.

### LINKED ISSUE #1387: Clients get 503s on server GracefulStop under heavy load.
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
v1.4.0
### What version of Go are you using (`go version`)?
1.8
### What operating system (Linux, Windows, …) and version?
Linux, Ubuntu 14.04.5 LTS
### What did you do?
If possible, provide a recipe for reproducing the error.
Our server serves @ 1.5 million RPM. We get 503s from our network proxy (envoy https://github.com/lyft/envoy) on a `GracefulStop` of the GRPC server.
We have a hot restart mechanism where we use `SO_REUSEPORT` to start a new server and drain the old server. The new server starts up fine and starts handling requests while clients of the old server on reports 503 (as the server is doing a graceful stop)
According to Matt Klein at Lyft @mattklein123 , the below could be a potential issue:
`There is a race condition inherent with GOAWAY and http/2. Basically, the GOAWAY can cross with new streams being sent. Those streams would then be reset by the server that sent GOAWAY. There is a workaround that people use (which Envoy does) which I'm sure Go is not doing. That workaround is basically to send 2 GOAWAY frames with a delay between them. The first GOAWAY has last stream ID set to max stream ID, after a delay, a real GOAWAY is sent.`

### What did you expect to see?
We expected to see a clean draining of the requests and no 5xx

### What did you see instead?
503s from the client.

### PR DESCRIPTION
**Pull request by MakMukhi**
Fixes #1387

### CODE REVIEW

File: transport/http2_server.go [hunk 1]
```diff
@@ -631,7 +638,7 @@ func (t *http2Server) handlePing(f *http2.PingFrame) {
 
 	if t.pingStrikes > maxPingStrikes {
 		// Send goaway and close the connection.
-		t.controlBuf.put(&goAway{code: http2.ErrCodeEnhanceYourCalm, debugData: []byte("too_many_pings")})
+		t.drain(http2.ErrCodeEnhanceYourCalm, []byte("too_many_pings"), true)
```
Comment (dfawley) [inline transport/http2_server.go]: Revert but add closeConn:true to the goAway.

File: transport/http2_server.go [hunk 2]
```diff
@@ -592,6 +595,10 @@ const (
 
 func (t *http2Server) handlePing(f *http2.PingFrame) {
 	if f.IsAck() {
+		if f.Data == goAwayPing.data {
```
Comment (dfawley) [inline transport/http2_server.go]: && t.drainChan != nil

File: transport/http2_server.go [hunk 3]
```diff
@@ -1116,7 +1155,17 @@ func (t *http2Server) RemoteAddr() net.Addr {
 }
 
 func (t *http2Server) Drain() {
-	t.controlBuf.put(&goAway{code: http2.ErrCodeNo})
+	t.drain(http2.ErrCodeNo, []byte{}, false)
+}
+
+func (t *http2Server) drain(code http2.ErrCode, debugData []byte, closeConn bool) {
```
Comment (dfawley) [inline transport/http2_server.go]: Remove closeConn

File: transport/http2_server.go [hunk 4]
```diff
@@ -1032,13 +1036,48 @@ func (t *http2Server) controller() {
...
+					// Follow that with a ping and wait for the ack to come back or a timer
+					// to expire. During this time accept new streams since they might have
+					// originated before the GoAway reaches the client.
+					// After getting the ack or timer expiration send out another GoAway this
+					// time with an ID of the max stream server intends to process.
+					t.framer.writeGoAway(true, math.MaxUint32, http2.ErrCodeNo, []byte{})
+					t.framer.writePing(true, false, goAwayPing.data)
+					go func() {
+						timer := time.NewTimer(time.Minute)
+						defer func() {
```
Comment (dfawley) [inline transport/http2_server.go]: defer timer.Stop()

File: transport/http2_server.go [hunk 5]
```diff
@@ -1032,13 +1036,48 @@ func (t *http2Server) controller() {
 						// The transport is closing.
 						return
 					}
+					ch := t.drainChan
```
Comment (dfawley) [inline transport/http2_server.go]: Remove local var

File: transport/http2_server.go [hunk 6]
```diff
@@ -1032,13 +1036,48 @@ func (t *http2Server) controller() {
...
 					sid := t.maxStreamID
-					t.state = draining
+					if i.isSecond {
```
Comment (dfawley) [inline transport/http2_server.go]: isSecond -> isFirst and reorganize?

File: transport/http2_server.go [hunk 7]
```diff
@@ -1032,13 +1036,48 @@ func (t *http2Server) controller() {
...
+						timer := time.NewTimer(time.Minute)
+						defer func() {
+							if !timer.Stop() {
+								<-timer.C
+							}
+						}()
+						select {
+						case <-ch:
+						case <-timer.C:
+							timer.Reset(infinity)
```
Comment (dfawley) [inline transport/http2_server.go]: Unnecessary Reset.

File: transport/http2_server.go [hunk 8]
```diff
@@ -1055,12 +1060,38 @@ func (t *http2Server) controller() {
 						return
 					}
 					sid := t.maxStreamID
-					t.state = draining
-					t.mu.Unlock()
-					t.framer.writeGoAway(true, sid, i.code, i.debugData)
-					if i.code == http2.ErrCodeEnhanceYourCalm {
-						t.Close()
+					if !i.headsUp {
+						// Stop accepting more stream now.
```
Comment (dfawley) [inline transport/http2_server.go]: stream_s_

**Comment by gitsen** [comment]
Thanks @MakMukhi ! Please drop us a note when you release this.

**Comment by menghanl** [comment]
@gitsen This was released in v1.5.2.

**Comment by gitsen** [comment]
Thanks @menghanl we will update and report our findings here

**Pull request merged**